sklearn/metrics/classification.py

```diff
@@ -391,72 +392,262 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+            meaningful for multilabel classification).
+
     normalize : bool, optional (default=True)
         If ``False``, return the sum of the Jaccard similarity coefficient
         over the sample set. Otherwise, return the average of Jaccard
-        similarity coefficient.
+        similarity coefficient. ``normalize`` is only meaningful when
+        ``average='samples'``.
 
     sample_weight : array-like of shape = [n_samples], optional
         Sample weights.
 
     Returns
     -------
-    score : float
-        If ``normalize == True``, return the average Jaccard similarity
-        coefficient, else it returns the sum of the Jaccard similarity
-        coefficient over the sample set.
-
-        The best performance is 1 with ``normalize == True`` and the number
-        of samples with ``normalize == False``.
+    score: float (if average is not None) or array of floats, shape =\
+            [n_unique_labels]
 
     See also
     --------
     accuracy_score, hamming_loss, zero_one_loss
 
-    Notes
-    -----
-    In binary and multiclass classification, this function is equivalent
-    to the ``accuracy_score``. It differs in the multilabel classification
-    problem.
-
     References
     ----------
     .. [1] `Wikipedia entry for the Jaccard index
            <https://en.wikipedia.org/wiki/Jaccard_index>`_
 
-
     Examples
     --------
     >>> import numpy as np
     >>> from sklearn.metrics import jaccard_similarity_score
-    >>> y_pred = [0, 2, 1, 3]
-    >>> y_true = [0, 1, 2, 3]
+
+    In the multilabel case with binary label indicators:
+
+    >>> y_true = np.array([[1, 0, 1], [0, 0, 1], [1, 1, 1]])
+    >>> y_pred = np.array([[0, 1, 1], [1, 1, 1], [0, 0, 1]])
     >>> jaccard_similarity_score(y_true, y_pred)
+    ... # doctest: +ELLIPSIS
+    0.33...
+    >>> jaccard_similarity_score(y_true, y_pred, average='micro')
+    ... # doctest: +ELLIPSIS
+    0.33...
+    >>> jaccard_similarity_score(y_true, y_pred, average='weighted')
     0.5
-    >>> jaccard_similarity_score(y_true, y_pred, normalize=False)
-    2
+    >>> jaccard_similarity_score(y_true, y_pred, average=None)
+    array([ 0.,  0.,  1.])
 
-    In the multilabel case with binary label indicators:
+    In the multiclass case:
 
-    >>> jaccard_similarity_score(np.array([[0, 1], [1, 1]]),\
-        np.ones((2, 2)))
-    0.75
+    >>> jaccard_similarity_score(np.array([0, 2, 1, 3]),
+    ... np.array([0, 1, 2, 3]), average='macro')
+    0.5
     """
 
+    average_options = (None, 'micro', 'macro', 'weighted', 'samples')
+    if average not in average_options and average != 'binary':
+        raise ValueError("average has to be one of " + str(average_options))
+
     # Compute accuracy for each possible representation
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     check_consistent_length(y_true, y_pred, sample_weight)
+    present_labels = unique_labels(y_true, y_pred)
```

Comment:
I won't force it, but I feel like some of this could be refactored with respect to precision, recal... (Indeed, all metrics could be calculated in one function if we wish)